docs/chymyst07.md

```diff
@@ -0,0 +1,731 @@
+
+TODO
+
+## Control over a shared resource (mutex, multiplex)
+
+### Single access
+
+Suppose we have an application with many concurrent processes and a shared resource _R_ (say, a database server) that should be only used by one process at a time.
+Let us assume that there is a certain function `doWork()` that will use the resource _R_ when called.
+Our goal is to make sure that `doWork()` is only called by at most one concurrent process at any time.
+While `doWork()` is being evaluated, we consider that the resource _R_ is not available.
+If `doWork()` is already being called by one process, all other processes trying to call `doWork()` should be blocked until the first `doWork()` is finished and the resource _R_ is again available.
+
+How would we solve this problem using the chemical machine?
+Since our only way to control concurrency is by manipulating molecules, we need to organize the chemistry such that `doWork()` is only called when certain molecules are available.
+In other words, `doWork()` must be called by a _reaction_ that consumes certain molecules whose presence or absence we will control.
+The reaction must be of the form `case [some molecules] => ... doWork() ...`.
+Let us call it the "worker reaction".
+There must be no other way to call `doWork()` except by starting this reaction.
+
+Processes that need to call `doWork()` will therefore need to emit a certain molecule that the worker reaction consumes.
+Let us call that molecule `request`.
+The `request` molecule must be a _blocking_ molecule because `doWork()` should be able to block the caller when the resource _R_ is not available.
+
+If the `request` molecule is the only input molecule of the worker reaction, we will be unable to prevent the reaction from starting whenever some process emits a `request` molecule.
+Therefore, we need a second input molecule in the worker reaction.
+Let us call that molecule `access`.
+The worker reaction will then look like this:
+
+```scala
+go { case access(_) + request(_, r) => ... doWork() ... }
+
+```
+
+Suppose some process emits a `request` molecule, and suppose this is the only process that does so at the moment.
+We should then allow the worker reaction to start and to complete `doWork()`.
+Therefore, `access()` must be present at the reaction site: we should have emitted it beforehand.
+
+While `doWork()` is evaluated, the `access()` molecule is absent, so no other copy of the worker reaction can start.
+This is precisely the exclusion behavior we need.
+
+When the `doWork()` function finishes, we need to unblock the calling process by replying to the `request` molecule.
+Perhaps `doWork()` will return a result value: in that case, we should pass this value as the reply value.
+We should also emit the `access()` molecule back into the soup, so that another process will be able to run `doWork`.
```

Comment:
similar to my smokers emitting themselves when they're done smoking and ready to get back to work at assembly line